File: solutions/go/sieve/1/sieve.go
package sieve

import "math"

func Sieve(limit int) []int {
    candidates := make(map[int]bool)

    for i := 2; i <= limit; i++ {
        candidates[i] = true
    }

    squareRoot := int(math.Ceil(math.Sqrt(float64(limit))))

    for i := 2; i <= squareRoot; i++ {
        if !candidates[i] {
            continue
        }

        for j := i * i; j <= limit; j += i {
            candidates[j] = false
        }
    }

    primes := []int{}

    for i := 2; i <= limit; i++ {
        if candidates[i] {
            primes = append(primes, i)
        }
    }

    return primes
}

File: solutions/ruby/sieve/1/sieve.rb
class Sieve
  attr_reader   :limit, :numbers
  attr_accessor :composites

  START = 2

  def initialize(limit)
    @limit      = limit
    @numbers    = (START..limit).to_a
    @composites = []
  end

  def primes
    numbers
      .each   { |n| collect_composites(n)   }
      .select { |n| !composites.include?(n) }
  end

  private

  def collect_composites(number)
    numbers
      .select { |n| n * number <= limit      }
      .each   { |n| composites << n * number }
  end
end
